compiler/noirc_evaluator/src/ssa/opt/inlining.rs

```diff
@@ -160,67 +188,94 @@ fn called_functions_vec(func: &Function) -> Vec<FunctionId> {
-fn get_functions_to_inline_into(
+///
+/// The returned `InlineInfos` won't have every function in it, only the ones which the algorithm visited.
+pub(super) fn compute_inline_infos(
     ssa: &Ssa,
     inline_no_predicates_functions: bool,
     aggressiveness: i64,
-) -> BTreeSet<FunctionId> {
-    let mut brillig_entry_points = BTreeSet::default();
-    let mut acir_entry_points = BTreeSet::default();
-
-    if matches!(ssa.main().runtime(), RuntimeType::Brillig(_)) {
-        brillig_entry_points.insert(ssa.main_id);
-    } else {
-        acir_entry_points.insert(ssa.main_id);
-    }
+) -> InlineInfos {
+    let mut inline_infos = InlineInfos::default();
+
+    inline_infos.insert(
+        ssa.main_id,
+        InlineInfo {
+            is_acir_entry_point: ssa.main().runtime().is_acir(),
+            is_brillig_entry_point: ssa.main().runtime().is_brillig(),
+            ..Default::default()
+        },
+    );
 
+    // Handle ACIR functions.
     for (func_id, function) in ssa.functions.iter() {
-        if matches!(function.runtime(), RuntimeType::Brillig(_)) {
+        if function.runtime().is_brillig() {
             continue;
         }
 
         // If we have not already finished the flattening pass, functions marked
         // to not have predicates should be preserved.
         let preserve_function = !inline_no_predicates_functions && function.is_no_predicates();
         if function.runtime().is_entry_point() || preserve_function {
-            acir_entry_points.insert(*func_id);
+            inline_infos.entry(*func_id).or_default().is_acir_entry_point = true;
         }
 
-        for called_function_id in called_functions(function) {
-            if matches!(ssa.functions[&called_function_id].runtime(), RuntimeType::Brillig(_)) {
-                brillig_entry_points.insert(called_function_id);
+        // Any Brillig function called from ACIR is an entry into the Brillig VM.
+        for called_func_id in called_functions(function) {
+            if ssa.functions[&called_func_id].runtime().is_brillig() {
+                inline_infos.entry(called_func_id).or_default().is_brillig_entry_point = true;
             }
         }
     }
 
     let times_called = compute_times_called(ssa);
 
-    let brillig_functions_to_retain: BTreeSet<_> = compute_functions_to_retain(
+    mark_brillig_functions_to_retain(
         ssa,
-        &brillig_entry_points,
-        &times_called,
         inline_no_predicates_functions,
         aggressiveness,
+        &times_called,
+        &mut inline_infos,
     );
 
-    acir_entry_points
-        .into_iter()
-        .chain(brillig_entry_points)
-        .chain(brillig_functions_to_retain)
-        .collect()
+    inline_infos
 }
 
-fn compute_times_called(ssa: &Ssa) -> HashMap<FunctionId, usize> {
+/// Compute the time each function is called from any other function.
+pub(super) fn compute_times_called(ssa: &Ssa) -> HashMap<FunctionId, usize> {
```

Comment:
Could we simplify this method using `compute_callers`? Maybe the indirection isn't worth it though, just a thought as I saw this method and `compute_callers` being called separately in `compute_bottom_up_order`. Ideally we could compute the callers and the total times called a single time.

Reply:
I think you're right, it can be calculated from that. I added the call count later to be able to roll up the weight as I put functions in place, and didn't want to touch what already did just what I needed, but if we can save an extra loop through everything, why not!

Reply:
Done 👍